src/reconcile/rec_visibility.c

```diff
@@ -869,26 +978,80 @@ __rec_fill_tw_from_upd_select(
      * that the value is visible to any timestamp/transaction id ahead of it.
      */
     if (upd->type == WT_UPDATE_TOMBSTONE) {
-        WT_TIME_WINDOW_SET_STOP(select_tw, upd, upd->prepare_state);
+        WT_TIME_WINDOW_SET_STOP(select_tw, upd, write_prepare);
         tombstone = upd_select->tombstone = upd;
 
-        /* Find the update this tombstone applies to. */
-        if (!__wt_txn_upd_visible_all(session, upd)) {
-            while (upd->next != NULL && upd->next->txnid == WT_TXN_ABORTED)
-                upd = upd->next;
+        /*
+         * Find the update this tombstone applies to.
+         *
+         * We need to find the full update if the prepared tombstone is rolled back. We resolve
+         * prepare updates recursively from the oldest to the newest. If we write a prepared
+         * tombstone, we may see a prepared update that is rolled back from the same transaction.
+         * Ensure we also write it to the disk in this case.
+         */
+        if (write_prepare || !__wt_txn_upd_visible_all(session, upd)) {
+            for (; upd->next != NULL; upd = upd->next) {
+                uint64_t next_txnid;
+                WT_ACQUIRE_READ(next_txnid, upd->next->txnid);
+                if (next_txnid != WT_TXN_ABORTED)
+                    break;
 
-            WT_ASSERT(session, upd->next == NULL || upd->next->txnid != WT_TXN_ABORTED);
+                if (!write_prepare)
+                    continue;
+
+                if (!F_ISSET(S2C(session), WT_CONN_PRESERVE_PREPARED))
+                    continue;
+
+                /* We may see aborted reserve updates in between the prepared updates. */
+                if (upd->next->type == WT_UPDATE_RESERVE)
+                    continue;
+
+                if (upd->next->prepare_state != WT_PREPARE_INPROGRESS)
+                    continue;
+
+                /*
+                 * Since we resolve prepared update from the oldest to newest, we may see a
+                 * tombstone still in prepared state but a prepared update that is rolled back from
+                 * the same transaction here. Handle this case.
+                 */
+                WT_ACQUIRE_READ(txnid, tombstone->txnid);
+                if (txnid == WT_TXN_ABORTED)
+                    txnid = tombstone->upd_saved_txnid;
+                if (upd->next->upd_saved_txnid == txnid) {
+                    WT_ASSERT(session, upd->next->prepare_ts == tombstone->prepare_ts);
+                    break;
+                }
+            }
+
+            WT_ASSERT(session,
+              upd->next == NULL || upd->next->txnid != WT_TXN_ABORTED ||
```

Comment:
I think this assert may also fire if we race. We need an acquire read for this upd->next->txnid != WT_TXN_ABORTED

Reply:
honestly i would remove this check `upd->next == NULL || upd->next->txnid != WT_TXN_ABORTED` and only check `(write_prepare && upd->next->prepare_state == WT_PREPARE_INPROGRESS && upd->next->upd_saved_txnid == tombstone->upd_saved_txnid)` if ` upd->next` is not null